### ISSUE TITLE
Version 60 performance 

### ISSUE THREAD

**Issue by RubenKelevra** [streetcomplete/StreetComplete#5994]
**How to Reproduce**

**Expected Behavior**

**Versions affected**

**Comment by westnordost** [comment]
Hm, I am going to need more information on that to be actionable.

My expectation is that Compose *is* slower than the normal Android View System, but it should of course stay on a non-noticeable level. (I, for one, didn't notice it. What's your device?)

Things to read: https://developer.android.com/develop/ui/compose/performance

**Comment by RubenKelevra** [comment]
Yeah I'm just unsure how to provide better data on this, as there's e.g. no FPS counter available on my device and recording screen captures is just 30 fps and robs performance anyway, so I doubt it will show up.


I'm using the Redmi Note 10 Pro, which is running Android 12 and I'm using 120 fps.

**Comment by RubenKelevra** [comment]
I've been scrolling around in SCEE and SC (to have the closes of comparison possible here between both versions - as I can't simply downgrade). I think the issue is limited to one thing: Loading of questmarkers which become visible while panning, zooming, rotation.

In 59.x it seems to be non-blocking, so the frames don't drop (much) while they are loaded and then smoothly transition in.

In SC 60 it looks like the renderer is blocked after the decision is made to load them, and will hang until they can be transitioned in, which is a choppy 2-3 frame animation.

Afterwards it's somewhat smooth again, meaning panning in a very small area works fine, panning a larger area will become very choppy as new quest markers need to be loaded in.

Hope this helps a bit?

**Comment by mnalis** [comment]
> In SC 60 it looks like the renderer is blocked after the decision is made to load them, and will hang until they can be transitioned in, which is a choppy 2-3 frame animation.

How choppy are we talking? if you are using 120fps, 2-3 frames drop would be temporary reduction to 40-60fps for a few hundredths of a second? I'm not sure if I'd be able to notice that :thinking: 

I have not noticed any difference on Samsung Galaxy S23+ (Android 13). Here is video (I've enabled `Show refresh rate` enabled in developer options and it showed `120` all the time; but that overlay isn't captuted by video recorder). It looks as smooth in real life as it does in the video (or maybe I'm so slow that I don't notice :smiley: )

https://github.com/user-attachments/assets/f25588cc-52e7-42d0-9d23-b0807180f56b

Should've I noticed issues here if I were affected by the problem (i.e. is it big enough area)? 
Does `Settings` / `Delete cache` help (after the data is repopulated, of course) ?

**Comment by westnordost** [comment]
Can you turn on location, @mnalis , and try again? (And compare with SCEE 59.x) Displaying this location pointer pin is a suspect for being performance hungry

**Comment by mnalis** [comment]
SCEE 59.0-ish:

https://github.com/user-attachments/assets/3bfcd1a5-5720-49fb-955a-3b68b950d88f

StreetComplete v60.0-alpha1:

https://github.com/user-attachments/assets/761d5c6f-4e27-4e99-a3e0-139870629a04

Judge by yourself. I personally do not seem to notice much difference between them on that Samsung Galaxy S23+ (Android 13), but than again, that phone is quite fast, and my perception of speed is not what it used to be anyway - it all seems zipping around at light speed to me, faster than my eyes can track it  :man_shrugging:

**Comment by RubenKelevra** [comment]
Well, your recordings show exactly the same issue I'm seeing, just my phone is probably 4 times slower, if not more ;)

When you pan large distances the motion freezes for a split second one or two times per panning. 

Here's a section of your video slowed down. The second pan got two of those large hangs:

https://github.com/user-attachments/assets/8a79058c-dbca-4f56-b15f-3f3138af8069

This may be a bit exaggerated, but it's basically what I'm seeing on my phone.

**Comment by westnordost** [comment]
Hmm, the map (and display plus loading of quests) hasn't been changed in v60, so maybe something else is taking more time so that there is less time on the main UI thread to render everything smoothly.

**Comment by matkoniecz** [comment]
Or it is case of Xiaomi phone being xiaomi-quality again? Mine xiaomi device at some point also started random hiccups with some of apps.

I will try to test it, maybe it is also affecting my phone, from the same manufacturer.

**Comment by westnordost** [comment]
Performance-wise, you can't compare a Samsung S23 with some low- or midrange phone.

**Comment by mnalis** [comment]
> Performance-wise, you can't compare a Samsung S23 with some low- or midrange phone.

I know (and have indicated as much), but that is what I have on me right now. Will be able to do comparison test in few days on older [Xiaomi Redmi note 6 Pro](https://www.gsmarena.com/xiaomi_redmi_note_6_pro-9333.php) which should be noticably less performant, if that is still of interest?

**Comment by westnordost** [comment]
Okay, I learned bit. So, the thing that is causing performance issues in Compose is always when for one reason or another the view recomposes more often than it needs to. It is key to find which parts cause the automatic recomposition. A few learnings:

1. Since Kotlin 2.0.20, some feature called [**strong skipping**](https://developer.android.com/develop/ui/compose/performance/stability/strongskipping) is enabled by default, which might speed up some things on recomposition. I didn't dive in further, I just upgraded Kotlin compiler from 2.0.0 to 2.0.21.

2. **Compose does not consider read-only collections immutable** (`List`, `Set`, etc.). Composables that take such collections as parameters _always_ recompose, because the data *might* have changed. These parameters should be marked as immutable explicitly, best by using the `Immutable*` classes from [`[kotlinx.collections.immutable](org.jetbrains.kotlinx:kotlinx-collections-immutable)`](https://github.com/Kotlin/kotlinx.collections.immutable).
  For example in the links screen, achievements screen, logs screen, quest selection screen, quest preset selection screen, edit statistics screen, we all have `List<SomeDataClass>` as parameter of composables. So, there is room for performance improvement.
  *However* on testing whether it makes a difference, I didn't find any difference. In fact, e.g. the links list does *not* recompose on e.g. scrolling. So maybe the `LazyColumn` etc. composables have something built in that already prevents recomposition? In any case, it doesn't seem to be the source of the reported performance problem. I will keep the `Immutable*` in mind in the future but it doesn't look like using that stuff is necessary right now

3. one can use `derivedState` to save on a few recompositions on edge cases. (E.g. don't need to decide whether to show the messages button at all each time the number of messages changes, only when the number of messages changes between zero and greater than zero)

4. Regarding debuggability, the [LayoutInspector](https://developer.android.com/develop/ui/compose/performance/tooling#layout-inspector) counts which parts of the view recompose how often, hence give an insight into what may cause performance issues. The composition tracing, mentioned on the same docs page, wasn't really insightful for me at all. Furthermore, when one sets breakpoints in composable functions, one can see *what exactly triggered this recompose*. **Very** helpful! ([since Android Studio Hedgehog](https://developer.android.com/studio/releases/past-releases/as-hedgehog-release-notes#compose-state-in-debugger)) 

Curiously, amongst others, the messages button and the upload button cause recompositions on every frame (i.e. whenever the map is moved) **because the function that should be called when they are clicked changes on every recomposition**. Will have to look into that. Probably it is not a good idea to define compose-function-local functions, as they seem to be re-created on every function call. (I thought they would be static or something)

**Comment by westnordost** [comment]
A good resource: https://multithreaded.stitchfix.com/blog/2022/08/05/jetpack-compose-recomposition/#gotcha---unstable-lambdas

**Comment by westnordost** [comment]
Actually, the gotcha linked is somewhat mitigated by the [strong skipping](https://developer.android.com/develop/ui/compose/performance/stability/strongskipping) feature. Which is good, because the solutions described in the above link make the code quite unreadable, IMO. I.e. the strong skipping does amongst others automatically what one would otherwise have to do manually in the own code: wrap the lambdas into `remember`

So, I only marked the ViewModels that are stable as with `@Stable` which should allow the Compose runtime to skip more recompositions.

Regarding the location pointer pin: The calculation of its position takes about 0.15 ms on my phone (upper-midrange phone from 2019). So, it's fast enough that I don't think it is worth checking whether to offload that into a background thread.

**Comment by FloEdelmann** [comment]
Ref:

* https://github.com/streetcomplete/StreetComplete/commit/7be0fcedfa3acd76b756e48c1d1d3819d0715254
* https://github.com/streetcomplete/StreetComplete/commit/b7de47f2162f4c41aba75ce8bf59168554ce09f5
* https://github.com/streetcomplete/StreetComplete/commit/3bfe25d0ae302a820ae988bb3eac1de776aee09b
* https://github.com/streetcomplete/StreetComplete/commit/05ebc526c4d7dbf5ce98893b447ab082711f4017
